### PR TITLE
Added option to change allocation range of published ports

### PR DESCRIPTION
**Pull request by suwang48404**
published-dynamic-port-range may be specified in command line 
dockerd ... -- published-dynamic-port-range="50000-60000" 
or in /etc/docker/daemon.json
published-dynamic-port-range:"50000-60000" specify the range used by
docker engine to allocate published service ports.

**- What I did**
Added published-dynamic-port-range option to allow user to change published service port 
range,

**- How I did it**
works in libnetwork and docker engine

**- How to verify it**
add/change published-dynamic-port-range is /etc/docker/daemon.json, and kill -SIGHUP DOCKERD_PID, and "docker run -p 80 SOME_WEB_SERVICE_IMG", the published port should 
be allocated from new range.

**- Description for the changelog**

Allow user to change range of allocated published ports.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by suwang48404** [comment]
@selansen @arkodg

**Comment by suwang48404** [comment]
@euanh @chiragtayal

File: cmd/dockerd/config.go [hunk 1]
```diff
@@ -85,6 +85,7 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) error {
 	flags.Var(opts.NewNamedListOptsRef("node-generic-resources", &conf.NodeGenericResources, opts.ValidateSingleGenericResource), "node-generic-resource", "Advertise user-defined resource")
 
 	flags.IntVar(&conf.NetworkControlPlaneMTU, "network-control-plane-mtu", config.DefaultNetworkMtu, "Network Control plane MTU")
+	flags.StringVar(&conf.HostServiceDynamicPortRange, "host-service-dynamic-port-range", "", "Port allocator range")
```
Comment (thaJeztah) [inline cmd/dockerd/config.go]: is this only affecting (swarm) services or also for `docker run -P` ? If the latter, we should pick a different name

File: daemon/daemon.go [hunk 2]
```diff
@@ -1410,6 +1410,10 @@ func (daemon *Daemon) networkOptions(dconfig *config.Config, pg plugingetter.Plu
 
 	options = append(options, nwconfig.OptionNetworkControlPlaneMTU(dconfig.NetworkControlPlaneMTU))
 
+	if len(dconfig.NetworkConfig.HostServiceDynamicPortRange) > 0 {
```
Comment (thaJeztah) [inline daemon/daemon.go]: I guess we should have some validation, as we likely want this to be within the regular ephemeral port range (ie, don't want port 80 to be wittin this range); also validate that the value is < 65535 (likely lower, to have a reasonable _range_)

File: cmd/dockerd/config.go (hunk 1 above)
Comment (thaJeztah) [reply cmd/dockerd/config.go]: Also, what is the format that this flag takes? Should this take a structured approach (`min=xx,max=xx`)?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: cmd/dockerd/config.go (hunk 1 above)
Comment (suwang48404) [reply cmd/dockerd/config.go]: yes, it will impact both swarm and docker run -P, will think of a better name

File: daemon/daemon.go (hunk 2 above)
Comment (suwang48404) [reply daemon/daemon.go]: the validation is in libnetwork

File: cmd/dockerd/config.go (hunk 1 above)
Comment (suwang48404) [reply cmd/dockerd/config.go]: The flag is --service-port-range "50000-60000"

**Comment by arkodg** [comment]
@suwang48404 would help if you shared some examples using the `moby` dev shell

**Comment by thaJeztah** [comment]
Some questions;

I think SwarmKit and "regular" containers use different ranges for the ephemeral port range; IIRC, regular containers use the range that's configured on the host (`net.ipv4.ip_local_port_range`), whereas SwarmKit (currently) has a fixed port range.

With this change;

- what is the default range? 
- is there something in place to prevent SwarmKit from allocating ports that were already used by "regular" containers (and vice-versa)?


Given that this affects both the ephemeral port range for `docker run`, and for swarm services (`docker service create` / `docker stack deploy`); when, and by "who", is the port selected? 

Given a cluster with (e.g.) 3 managers and several workers;

If I create a service with a port published through the ingress network;

- will the manager I'm connected to pick the port?
- will the worker? (likely not)

Same, but when creating a service with `host` mode publishing;

- will the manager I'm connected to pick the port?
- will the worker? (likely not)

Give that this is a configuration on the _daemon_, we must assume that each daemon can have a different configuration, which means that if the _manager_ picks the port, that the selected port range can be different depending on which manager I'm connected to.

So, should we store this configuration in the swarm raft store instead? (similar to what we do for `--default-addr-pool`, which is configured for the _cluster_)

### setting and updating this configuration

Per my earlier comment; validation is done in the libnetwork code; I still have to build and test this change, but;

- at what point is the given value validated?
- is the value validated when reloading the configuration (`SIGHUP`); in other words; do we have guards in place to prevent reloading with an invalid configuration?


(Without having given this much thought, so just thinking out loud); Are there possible complications if I update this port-range? What happens to existing containers if the daemon restarts? I assume containers (and services) preserve the port, but does the daemon / swarm continue to keep track of ports that were allocated in the old range?

**Comment by suwang48404** [comment]
Hi Sebastiaan, 

thx for review, and spending time think about it.

reply inline please

> Some questions;
> 
> I think SwarmKit and "regular" containers use different ranges for the ephemeral port range; IIRC, regular containers use the range that's configured on the host (`net.ipv4.ip_local_port_range`), whereas SwarmKit (currently) has a fixed port range.
> 
> With this change;
> 
> * what is the default range?
====> on linux based system, 49153-60999, on windows 60000-65000

> * is there something in place to prevent SwarmKit from allocating ports that were already used by "regular" containers (and vice-versa)?
> 
===> swarm allocating range is hard coded to 30000-32767, and they are not yet tunable

> Given that this affects both the ephemeral port range for `docker run`, and for swarm services (`docker service create` / `docker stack deploy`); when, and by "who", is the port selected?
> 
=====> docker engine port allocator is used when "docker run -p ..." and "docker service create  --publish mode=host", swarm port allocator is used when "docker service create --publish .."

> Given a cluster with (e.g.) 3 managers and several workers;
> 
> If I create a service with a port published through the ingress network;
> 
> * will the manager I'm connected to pick the port?
=====> if  "--publish mode" setting is not host, swarm manager pick a port from its port allocator, and this port is reserved on all nodes in cluster

> * will the worker? (likely not)
======> correct. docker engine will not allocate port

> 
> Same, but when creating a service with `host` mode publishing;
> 
> * will the manager I'm connected to pick the port?
====> swarm defers to docker engine to allocate port

> * will the worker? (likely not)
==========> docker engine allocate port if "docker service create --publish mode=host,...",the allocated port has only local node scope.

> 
> Give that this is a configuration on the _daemon_, we must assume that each daemon can have a different configuration, which means that if the _manager_ picks the port, that the selected port range can be different depending on which manager I'm connected to.
> 
> So, should we store this configuration in the swarm raft store instead? (similar to what we do for `--default-addr-pool`, which is configured for the _cluster_)
> 
========> each docker engine picks different port independently as port allocated by docker engine has local scope only, does not impact other node? That is, the same port allocated by docker engines can be allocated and assigned to same/different service instances on different nodes. Do we need raft in this use case?

> ### setting and updating this configuration
> Per my earlier comment; validation is done in the libnetwork code; I still have to build and test this change, but;
> 
> * at what point is the given value validated?
====> validation take place when config is read via daeomon.json or dockerd parameter

> * is the value validated when reloading the configuration (`SIGHUP`); in other words; do we have guards in place to prevent reloading with an invalid configuration?
> 
=========> yes

> (Without having given this much thought, so just thinking out loud); Are there possible complications if I update this port-range? What happens to existing containers if the daemon restarts? I assume containers (and services) preserve the port, but does the daemon / swarm continue to keep track of ports that were allocated in the old range?
=======> The follows refers to services instances created with "--publish mode=host",  or via "docker run -p", if port range is changed, config reloaded, existing container will continue to use old port allocated even if it is outside of new range; if dockerd rebooted, and containers restarted, they will be given ports from the new range. 
=======> for port allocated by swarm, i.e. "docker service create --publish ....", I believe port persists across docker engine restart.

Thyx, Su

**Comment by akerouanton** [comment]
For the record, the libnetwork changes needed by this PR have been introduced in moby/libnetwork#2461 and is being reverted by moby/moby#43066. If someone wants to work on this PR, they'd have to re-apply the reverted commit and fix the issue described in moby/moby#43054.

@thaJeztah There's been no activity on this PR since two years and there's no linked issue so I'm not sure if that's really something people want. Moreover, the port range can already by configured through sysctl, so this PR only brings the config reload part. Maybe this change should be closed and mark as "won't implement" (unless there're requests for it).

**Comment by tianon** [comment]
Agreed, IMO we should close this for now in favor of https://github.com/moby/moby/pull/43066.

**Comment by thaJeztah** [comment]
@neersighted @corhere @evol262 FYI - I recall the motivation for these changes was related to (at the time) Docker EE, to be able to reserve separate "ephemeral" port ranges for SwarmKit, "docker run", and k8s (?). Perhaps you're already carrying patches for that, but if not, perhaps this needs to be carried at some point

**Pull request closed without merging**